src/core/ext/transport/chttp2/transport/frame.cc

```diff
@@ -0,0 +1,519 @@
+  }
+
+  void operator()(Http2SettingsFrame& frame) {
+    const size_t payload_size = 6 * frame.settings.size();
+    auto hdr_and_payload =
+        extra_bytes_.TakeFirst(kFrameHeaderSize + payload_size);
+    Http2FrameHeader{static_cast<uint32_t>(payload_size), kFrameTypeSettings,
+                     MaybeFlag(frame.ack, kFlagAck), 0}
+        .Serialize(hdr_and_payload.begin());
+    size_t offset = kFrameHeaderSize;
+    for (auto& setting : frame.settings) {
+      Write2b(setting.id, hdr_and_payload.begin() + offset);
+      Write4b(setting.value, hdr_and_payload.begin() + offset + 2);
+      offset += 6;
+    }
+    out_.AppendIndexed(Slice(std::move(hdr_and_payload)));
+  }
+
+  void operator()(Http2PingFrame& frame) {
+    auto hdr_and_payload = extra_bytes_.TakeFirst(kFrameHeaderSize + 8);
+    Http2FrameHeader{8, kFrameTypePing, MaybeFlag(frame.ack, kFlagAck), 0}
+        .Serialize(hdr_and_payload.begin());
+    Write8b(frame.opaque, hdr_and_payload.begin() + kFrameHeaderSize);
+    out_.AppendIndexed(Slice(std::move(hdr_and_payload)));
+  }
+
+  void operator()(Http2GoawayFrame& frame) {
+    auto hdr_and_fixed_payload = extra_bytes_.TakeFirst(kFrameHeaderSize + 8);
+    Http2FrameHeader{static_cast<uint32_t>(8 + frame.debug_data.length()),
+                     kFrameTypeGoaway, 0, 0}
+        .Serialize(hdr_and_fixed_payload.begin());
+    Write4b(frame.last_stream_id,
+            hdr_and_fixed_payload.begin() + kFrameHeaderSize);
+    Write4b(frame.error_code,
+            hdr_and_fixed_payload.begin() + kFrameHeaderSize + 4);
+    out_.AppendIndexed(Slice(std::move(hdr_and_fixed_payload)));
+    out_.AppendIndexed(std::move(frame.debug_data));
+  }
+
+  void operator()(Http2WindowUpdateFrame& frame) {
+    auto hdr_and_payload = extra_bytes_.TakeFirst(kFrameHeaderSize + 4);
+    Http2FrameHeader{4, kFrameTypeWindowUpdate, 0, frame.stream_id}.Serialize(
+        hdr_and_payload.begin());
+    Write4b(frame.increment, hdr_and_payload.begin() + kFrameHeaderSize);
+    out_.AppendIndexed(Slice(std::move(hdr_and_payload)));
+  }
+
+  void operator()(Http2UnknownFrame&) { Crash("unreachable"); }
+
+ private:
+  SliceBuffer& out_;
+  MutableSlice extra_bytes_;
+};
+
+absl::Status StripPadding(SliceBuffer& payload) {
+  if (payload.Length() < 1) {
+    return absl::InternalError("padding flag set but no padding byte");
+  }
+  uint8_t padding_bytes;
+  payload.MoveFirstNBytesIntoBuffer(1, &padding_bytes);
+  if (payload.Length() < padding_bytes) {
+    return absl::InternalError("padding flag set but not enough padding bytes");
+  }
+  payload.RemoveLastNBytes(padding_bytes);
+  return absl::OkStatus();
+}
+
+absl::StatusOr<Http2DataFrame> ParseDataFrame(const Http2FrameHeader& hdr,
+                                              SliceBuffer& payload) {
+  if (hdr.flags & ~(kFlagEndStream | kFlagPadded)) {
+    return absl::InternalError(
+        absl::StrCat("unsupported data flags: ", hdr.ToString()));
```

Comment:
As per RFC 9113, `Unused flags MUST be ignored on receipt`

Reply:
fixed